Issue #6201: Add support for Spanner Postgres

code-ape opened the pull request:
## Related PRs and Issues

1. #6201

## Reason 

This PR addresses issue #6201 by adding a dedicated dialect for Google Cloud Spanner's flavor of "Postgres". This is because Spanner's Postgres implementation (via `pgadapter`) is a limited subset of the full Postgres dialect and thus minor changes are needed.

## Todo

- [ ] Code changes
- [ ] Add tests
- [ ] Update docs

## Changes

1. Added dialect `postgres-spanner` via `lib/constants.js`, `lib/dialects/index.js`, and `lib/dialects/postgres-spanner/`.
2. Added Docker Compose entries for the local Spanner emulator and PgAdapter which proxies the Postgres dialect requests, via `scripts/docker-compose.yml`.

## References
1. https://cloud.google.com/spanner/docs/postgresql-interface
2. https://cloud.google.com/spanner/docs/pgadapter

code-ape committed:
Issue #6201: Add support for Spanner Postgres

code-ape committed:
Fixing formatting

kibertoad commented on lib/dialects/index.ts:
```diff
@@ -10,6 +10,7 @@ const dbNameToDialectLoader: Record<string, () => any> = Object.freeze({
   oracledb: () => require('./oracledb'),
   pgnative: () => require('./pgnative'),
   postgres: () => require('./postgres'),
+  'postgres-spanner': () => require('./postgres-spanner'),
```
can you also add it to documentation?

kibertoad commented on scripts/docker-compose.yml:
```diff
@@ -84,6 +84,34 @@ services:
       - -c
       - 'until /usr/local/bin/psql postgres://testuser:knextest@postgres/knex_test -c "SELECT 1"; do sleep 5; done'
 
+  spanner-emulator:
```
can you configure integration tests to run on this?

(probably not all of them would be passing, and it's ok to skip the ones that are hard to fix immediately, but the more coverage we can get, the better)

OlivierCavadenti committed:
feat(postgres-spanner): fix dialect wiring, add integration tests and docs

- Set dialect/driverName on Client_PgSpanner's own prototype instead of
  mutating Client_PG's (the latter broke every postgres client).
- Register postgres-spanner in the test knexfile and the CI integration matrix.
- Rename the pgadapter compose service to postgres-spanner and add a wait
  step that bootstraps the knex_test database on the Spanner emulator.
- Document the postgres-spanner client and its limitations (no schemas).

OlivierCavadenti committed:
test(postgres-spanner): register dialect in integration test provider

getKnexForDb looked up postgres-spanner in the integration2 provider's
own testConfigs map (separate from test/knexfile.js) and got undefined,
crashing the root before-all hook. Register the dialect there too.

OlivierCavadenti committed:
test(postgres-spanner): run a dedicated integration spec on the emulator

Spanner's PostgreSQL interface is a limited subset (no schemas, stricter
typing, migrations unsupported), so it can't run the shared integration
suite. Instead run a focused spec against the Spanner emulator covering the
supported surface and the dialect's compiler overrides (CRUD, where,
columnInfo, hasTable, hasColumn), wired via a dedicated mocha config and a
separate CI step.

OlivierCavadenti committed:
fix(postgres-spanner): columnInfo without current_database()/current_schema()

Spanner's PostgreSQL interface supports neither current_database() nor
current_schema(), which the base PostgreSQL columnInfo query relies on.
Override columnInfo to drop the table_catalog filter.

OlivierCavadenti committed:
chore(postgres-spanner): trim verbose code comments

OlivierCavadenti commented:
Picked this up and rebased on master. 

Fixed two things: the dialect mutated the base `pg` client prototype (broke all postgres), and `columnInfo` used `current_database()`/`current_schema()` which Spanner doesn't support.

The shared integration suite assumes schemas and migrations that Spanner lacks (~133 failures in the common setup), so I added a dedicated spec running on the Spanner emulator: CRUD, `where`, `columnInfo`, `hasTable`/`hasColumn`. Docs updated, CI green.

n1ghtmare commented:
Thank you @OlivierCavadenti for picking this up, much appreciated. 

One thing that this PR doesn't cover yet that we've had to patch downstream, is migration rollback ordering:

https://github.com/knex/knex/blob/8355ad758db9ebc00aacdb8ca6f6d56f28bb5f4a/lib/migrations/migrate/Migrator.js#L602-L617

`_getLastBatch()` orders the last batch by `id DESC`, which assumes the auto-increment `id` ascends with insertion order. That doesn't hold up on Spanner. The migrations table is created with `t.increments()`, which on the Postgres dialiect compiles to `SERIAL` - and per the spanner docs, [`SERIAL`/`AUTO_INCREMENT` are  DDL aliases for `IDENTITY` columns](https://cloud.google.com/spanner/docs/primary-key-default-value) backed by **bit-reversed sequences**, whose values are "evenly distributed across the key space so that they don't cause hotspots" rather than monotonic. This is by design - Spanner's [schema-design best practices](https://cloud.google.com/spanner/docs/schema-design) say *"Do not choose a column whose value monotonically increases or decreases as the first key part…"*

So on Spanner, `ORDER BY id DESC` does **not** return the last batch in reverse insertion order, and `migrate:rollback` can run the batch's `down()` steps out of order.

We patch this to order by `migration_time`. Suggested fix that's safe for **all** dialects (unless I'm missing something, sorry if that's the case):

  ```js
  .orderBy('migration_time', 'desc')
  .orderBy('id', 'desc');
  ```
  
(The `migration_time` tiebreaker matters because, I think, a fast batch can stamp several rows with the same `new Date();` the secondary `id` keeps ordering deterministic on dialects where it's monotonic, and is harmless on Spanner.)
  
  Since Spanner support is the goal of this PR, would it make sense to include this here?
